src/core/ddsi/src/ddsi_typelookup.c

```diff
@@ -0,0 +1,356 @@
+  const type_lookup_request_t *req = sample->sample;
+  GVTRACE (" handle-tl-req wr "PGUIDFMT " seqnr %"PRIi64" ntypeids %"PRIu32, PGUID (req->writer_guid), req->sequence_number, req->type_ids.n);
+
+  ddsrt_mutex_lock (&gv->tl_admin_lock);
+  type_identifier_type_object_pair_seq_t types = { 0, NULL };
+  for (uint32_t n = 0; n < req->type_ids.n; n++)
+  {
+    type_identifier_t *tid = &req->type_ids.type_ids[n];
+    GVTRACE (" type "PTYPEIDFMT, PTYPEID (*tid));
+    struct tl_meta *tlm = ddsi_tl_meta_lookup_locked (gv, tid);
+    if (tlm != NULL && tlm->state == TL_META_RESOLVED)
+    {
+      size_t sz;
+      types.n++;
+      types.types = ddsrt_realloc (types.types, types.n * sizeof (*types.types));
+      memcpy (&types.types[types.n - 1].type_identifier, tlm->type_id, sizeof (types.types[types.n - 1].type_identifier));
+      tlm->sertype->ops->serialize (tlm->sertype, &sz, &types.types[types.n - 1].type_object.value);
+      assert (sz <= UINT32_MAX);
+      types.types[types.n - 1].type_object.length = (uint32_t) sz;
+      GVTRACE (" found");
+    }
+  }
+  ddsrt_mutex_unlock (&gv->tl_admin_lock);
+
+  if (types.n > 0)
+  {
+    struct participant *pp = get_typelookup_pp (gv, guid_prefix);
+    assert (pp != NULL);
+    write_typelookup_reply (pp, req->sequence_number, &types);
+    for (uint32_t n = 0; n < types.n; n++)
+      ddsrt_free (types.types[n].type_object.value);
+    ddsrt_free (types.types);
+  }
+}
+
+void ddsi_tl_handle_reply (struct ddsi_domaingv *gv, struct ddsi_serdata *sample_common)
+{
+  struct generic_proxy_endpoint **gpe_match_upd = NULL;
+  uint32_t n_match_upd = 0;
+  assert (!(sample_common->statusinfo & (NN_STATUSINFO_DISPOSE | NN_STATUSINFO_UNREGISTER)));
+  const struct ddsi_serdata_pserop *sample = (const struct ddsi_serdata_pserop *) sample_common;
+  const type_lookup_reply_t *reply = sample->sample;
+  struct ddsi_sertype_default *st = NULL;
+  uint32_t n = 0;
+
+  GVTRACE ("handle-tl-reply wr "PGUIDFMT " seqnr %"PRIi64" ntypeids %"PRIu32, PGUID (reply->writer_guid), reply->sequence_number, reply->types.n);
+  while (n < reply->types.n)
+  {
+    ddsrt_mutex_lock (&gv->tl_admin_lock);
+    type_identifier_type_object_pair_t r = reply->types.types[n];
+    struct tl_meta *tlm = ddsi_tl_meta_lookup_locked (gv, &r.type_identifier);
+    if (tlm != NULL && tlm->state == TL_META_REQUESTED && tlm->n_endpoints > 0)
+    {
+      GVTRACE (" type "PTYPEIDFMT, PTYPEID (r.type_identifier));
+      st = ddsrt_malloc (sizeof (*st));
+      ddsi_sertype_init_from_ser (gv, &st->c, &ddsi_sertype_ops_default, r.type_object.length, r.type_object.value);
+      ddsrt_mutex_lock (&gv->sertypes_lock);
+      ddsi_sertype_register_locked (&st->c);
+      ddsrt_mutex_unlock (&gv->sertypes_lock);
+
+      tlm->state = TL_META_RESOLVED;
+      tlm->sertype = &st->c; // refcounted by register
+
+      gpe_match_upd = ddsrt_malloc (tlm->n_endpoints * sizeof (*gpe_match_upd));
```

Comment:
This, inside the outer loop, is not going to work quite so well. And I'm tempted to suggest adding some generic list handling (in `ddsrt`) to make this sort of thing a bit less verbose. Something that constructs a linked list of (short) arrays of entries or so. It could be templatised with macros, too 🙂

Reply:
I've created an abstraction (the `tlm_endpoints_*` functions) for the endpoint list in `struct tl_meta`, which is currently implemented as a simple array, but can easily be replaced by another list implementation if required.